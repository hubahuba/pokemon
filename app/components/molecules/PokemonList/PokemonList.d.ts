export interface PokemonListProps {
  testID?: string;
  data?: PokemonData[] | PokemonData[][];
  onEndReach?: () => void;
  onPressCard?: (id: PokemonData) => void;
}
