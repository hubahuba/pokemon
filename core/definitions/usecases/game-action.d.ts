import {BerryData, PokemonData} from '@/definitions/usecases/pokemon';

export interface GameActionUseCase {
  iChooseYou: (PokemonData) => PokemonData;
  deletePokemon: (pokemonId: string) => void;
  feedPokemon: (pokemonId: string, berry: BerryData) => PokemonData | undefined;
  evolutionPokemon: (pokemonId: string) => PokemonData | undefined;
}
