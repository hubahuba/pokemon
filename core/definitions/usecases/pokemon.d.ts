import {PokemonData} from '@/definitions/data/pokemon.definition';
import {BerryData} from '@/definitions/data/berries.definition';
import {UseInfiniteQueryResult} from '@tanstack/react-query';
import {
  Exact,
  InputMaybe,
  Pokemon_V2_Pokemon_Bool_Exp,
} from '@/definitions/gql/codegen/graphql';
import Pokemon from '@/usecases/pokemon.ts';

export type QuerySearch = Exact<{
  limit: number;
  offset: number;
  search?: InputMaybe<Pokemon_V2_Pokemon_Bool_Exp> | undefined;
}>;

export interface PokemonStat {
  value: number;
  name?: string;
}

export interface PokemonData {
  id: number;
  name: string;
  baseWeight?: number | null;
  currentWeight?: number | null;
  maxWeight?: number | null;
  nextEvolution?: PokemonData[];
  image: any;
  order?: number | null;
  stats?: PokemonStat[];
  evolutions?: PokemonData[];
  lastEat?: string;
  owned?: boolean;
  ownedId?: string;
}

export interface BerryData {
  id: number;
  name: string;
  firmness?: 'very-soft' | 'soft' | 'hard' | 'very-hard' | 'super-hard';
  image: any;
}

export interface BerryFirmness {
  'very-soft': number;
  soft: number;
  hard: number;
  'very-hard': number;
  'super-hard': number;
}

export interface GetBerriesUseCase {
  data: BerryData[] | undefined;
  isFetching: boolean;
  isFetched: boolean;
  error: Error | null;
}

export interface PokemonUseCaseProps {
  getPokemon: (
    offset: number,
    limit: number,
    search?: string | number,
  ) => UseInfiniteQueryResult<PokemonData[] | undefined>;

  getBerries: () => GetBerriesUseCase;
  getMyPokemon: () => PokemonData[];
}

export interface getBerryScoreResult {
  firmness: string;
  score: number;
}

export interface fetchPokemonResponse {
  lastOffset: number;
  pokemons: Pokemon[];
}
