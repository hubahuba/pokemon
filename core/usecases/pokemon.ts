import 'react-native-get-random-values';
import {
  GetBerriesUseCase,
  PokemonUseCaseProps,
  PokemonData,
  QuerySearch,
} from '@/definitions/usecases/pokemon';
import {pokeAPIQueryDocument} from '@/data/pokemons.ts';
import {PokeGetBerriesQueryDocument} from '@/data/berries.ts';
import {pokemonDataMapper, berryDataMapper, filterMapper} from '@/utils';
import {
  useInfiniteQuery,
  UseInfiniteQueryResult,
  useQuery,
} from '@tanstack/react-query';
import request from 'graphql-request';
import Config from 'react-native-config';
import LocalStorage from '@/adapters/local-storage.ts';

const Pokemon: PokemonUseCaseProps = {
  getPokemon: (
    offset,
    limit,
    search,
  ): UseInfiniteQueryResult<PokemonData[] | undefined> => {
    const filter: QuerySearch = filterMapper(offset, limit, search);

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const query = useInfiniteQuery({
      queryKey: ['getPokemons', filter],
      queryFn: async ({pageParam}: any) => {
        const resp = await request(
          Config.API_URL as string,
          pokeAPIQueryDocument,
          pageParam,
        );
        return {
          lastOffset: pageParam.offset,
          pokemons: pokemonDataMapper(resp),
        };
      },
      initialPageParam: filter,
      getNextPageParam: (nextPage, allPages, lastPageParam) => {
        if (lastPageParam && lastPageParam.limit) {
          lastPageParam.offset = nextPage.lastOffset + lastPageParam.limit;
          return lastPageParam;
        }
        return undefined;
      },
    });

    let pokemonData: PokemonData[] = [];
    if (query && query.data) {
      const pokemons = query.data.pages.map(group => group.pokemons);
      pokemonData = pokemons?.reduce((r, e) => (r.push(...e), r), []);
    }

    // @ts-ignore
    return {
      ...query,
      data: pokemonData,
    };
  },

  getBerries: (): GetBerriesUseCase => {
    const localBerries = LocalStorage.getItem('berries');
    if (localBerries) {
      const data = JSON.parse(localBerries);
      return {
        isFetched: true,
        isFetching: false,
        error: null,
        data,
      };
    }
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const {data, isFetching, isFetched, error} = useQuery({
      queryKey: ['getBerries'],
      queryFn: async () => {
        const resp = await request(
          Config.API_URL as string,
          PokeGetBerriesQueryDocument,
        );
        return berryDataMapper(resp);
      },
    });

    if (data) {
      LocalStorage.setItem('berries', JSON.stringify(data));
    }

    return {
      data,
      isFetching,
      isFetched,
      error,
    };
  },

  getMyPokemon: (): PokemonData[] => {
    const localBerries = LocalStorage.getItem('my-pokemons');
    if (localBerries) {
      return JSON.parse(localBerries);
    }
    return [];
  },
};

export default Pokemon;
