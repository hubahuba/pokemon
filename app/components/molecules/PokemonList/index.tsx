import * as React from 'react';
import {View, FlatList} from 'react-native';
import PokemonCard from '~/components/molecules/PokemonCard';
import {PokemonListProps} from './PokemonList';
import {PokemonData} from '@/definitions/usecases/pokemon';

function PokemonList({
  data,
  onEndReach,
  onPressCard,
}: Readonly<PokemonListProps>) {
  return (
    <View>
      <View className="flex flex-row flex-wrap justify-center items-start">
        <FlatList
          data={data as PokemonData[]}
          numColumns={2}
          renderItem={({item}) => (
            <PokemonCard
              onPress={() => onPressCard?.(item)}
              name={item.name}
              image={item.image}
              testID={`PokemonCard-${item.name}`}
            />
          )}
          keyExtractor={item => item.ownedId || String(item.id)}
          columnWrapperStyle={{
            justifyContent: 'center',
          }}
          onEndReachedThreshold={0.3}
          onEndReached={() => onEndReach?.()}
          maxToRenderPerBatch={20}
          removeClippedSubviews={true}
          testID="PokemonListContainer"
        />
      </View>
    </View>
  );
}

export default PokemonList;
