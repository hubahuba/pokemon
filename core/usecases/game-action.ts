import {GameActionUseCase} from '@/definitions/usecases/game-action';
import {PokemonData} from '@/definitions/usecases/pokemon';
import LocalStorage from '@/adapters/local-storage.ts';
import {v4 as uuid} from 'uuid';
import {findNextWeight, getBerryScore} from '@/utils';

const GameAction: GameActionUseCase = {
  iChooseYou: data => {
    const selected: PokemonData = {...data};
    let myPokemons: PokemonData[] | string | undefined =
      LocalStorage.getItem('my-pokemons');
    if (myPokemons) {
      myPokemons = JSON.parse(myPokemons);
    } else {
      myPokemons = [];
    }
    selected.owned = true;
    selected.ownedId = uuid();
    selected.currentWeight = selected.baseWeight;
    (myPokemons as PokemonData[]).push(selected);
    LocalStorage.setItem('my-pokemons', JSON.stringify(myPokemons));
    return selected;
  },

  deletePokemon: pokemonId => {
    const myPokemons: PokemonData[] | string | undefined =
      LocalStorage.getItem('my-pokemons');
    if (myPokemons) {
      const pokemons: PokemonData[] = JSON.parse(myPokemons);
      const filtered = pokemons.filter(
        pokemon => pokemon.ownedId !== pokemonId,
      );
      LocalStorage.setItem('my-pokemons', JSON.stringify(filtered));
    }
  },

  feedPokemon: (pokemonId, berry) => {
    const myPokemons: PokemonData[] | string | undefined =
      LocalStorage.getItem('my-pokemons');
    if (myPokemons) {
      const pokemons: PokemonData[] = JSON.parse(myPokemons);
      const currentPokemon = pokemons.find(
        pokemon => pokemon.ownedId === pokemonId,
      );
      if (currentPokemon) {
        const score = getBerryScore(
          berry,
          currentPokemon.lastEat === berry.firmness,
        );
        currentPokemon.lastEat = score.firmness;
        let weight = (currentPokemon.currentWeight as number) + score.score;
        if (weight < 0) {
          weight = 1;
        } else if (weight >= (currentPokemon.maxWeight as number)) {
          weight = currentPokemon.maxWeight as number;
        }
        currentPokemon.currentWeight = weight;
        const filtered = pokemons.filter(
          pokemon => pokemon.ownedId !== pokemonId,
        );
        filtered.push(currentPokemon);
        LocalStorage.setItem('my-pokemons', JSON.stringify(filtered));
        return currentPokemon;
      }
      return currentPokemon;
    }
    return undefined;
  },

  evolutionPokemon: pokemonId => {
    const myPokemons: PokemonData[] | string | undefined =
      LocalStorage.getItem('my-pokemons');
    if (myPokemons) {
      const pokemons: PokemonData[] = JSON.parse(myPokemons);
      let currentPokemon = pokemons.find(
        pokemon => pokemon.ownedId === pokemonId,
      );
      if (currentPokemon) {
        if (currentPokemon.nextEvolution) {
          const nextEvo = [...currentPokemon.nextEvolution];
          const current = nextEvo.shift();
          current.owned = currentPokemon.owned;
          current.ownedId = currentPokemon.ownedId;
          const nextOfNextEvo = findNextWeight(current, nextEvo);
          if (nextOfNextEvo.length > 0 && nextOfNextEvo[0].id !== current.id) {
            current.nextEvolution = nextOfNextEvo;
            current.maxWeight = nextOfNextEvo[0].baseWeight;
          } else {
            current.maxWeight = current.baseWeight;
          }

          const filtered = pokemons.filter(
            pokemon => pokemon.ownedId !== pokemonId,
          );
          filtered.push(current);
          LocalStorage.setItem('my-pokemons', JSON.stringify(filtered));

          return current;
        }

        return currentPokemon;
      }
      return currentPokemon;
    }
    return undefined;
  },
};

export default GameAction;
