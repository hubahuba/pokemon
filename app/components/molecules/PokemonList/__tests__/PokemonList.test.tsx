import React from 'react';
import renderer from 'react-test-renderer';
import PokemonList from '../index';
import {PokemonData} from '@/definitions/usecases/pokemon';

it('renders Card correctly', () => {
  const dummyPokemon: PokemonData[] = [
    {
      id: 1,
      name: 'pokemon',
      baseWeight: 10,
      currentWeight: 10,
      maxWeight: 10,
      nextEvolution: [],
      image: 'poke-image',
      order: 1,
      stats: [
        {
          name: 'stat',
          value: 12,
        },
      ],
      evolutions: [],
      lastEat: 'soft',
      owned: false,
    },
  ];

  const tree = renderer.create(<PokemonList data={dummyPokemon} />).toJSON();
  expect(tree).toMatchSnapshot();
});
