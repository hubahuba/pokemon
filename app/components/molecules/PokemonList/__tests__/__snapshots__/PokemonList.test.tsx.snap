// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Card correctly 1`] = `
<View>
  <View
    style={
      [
        {
          "display": "flex",
        },
        {
          "flexDirection": "row",
        },
        {
          "flexWrap": "wrap",
        },
        {
          "justifyContent": "center",
        },
        {
          "alignItems": "flex-start",
        },
      ]
    }
  >
    <RCTScrollView
      data={
        [
          {
            "baseWeight": 10,
            "currentWeight": 10,
            "evolutions": [],
            "id": 1,
            "image": "poke-image",
            "lastEat": "soft",
            "maxWeight": 10,
            "name": "pokemon",
            "nextEvolution": [],
            "order": 1,
            "owned": false,
            "stats": [
              {
                "name": "stat",
                "value": 12,
              },
            ],
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      maxToRenderPerBatch={20}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onEndReachedThreshold={0.3}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={true}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              [
                {
                  "flexDirection": "row",
                },
                {
                  "justifyContent": "center",
                },
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "borderBottomColor": "#d1d5db",
                  "borderBottomLeftRadius": 4,
                  "borderBottomRightRadius": 4,
                  "borderBottomWidth": 1,
                  "borderLeftColor": "#d1d5db",
                  "borderLeftWidth": 1,
                  "borderRightColor": "#d1d5db",
                  "borderRightWidth": 1,
                  "borderTopColor": "#d1d5db",
                  "borderTopLeftRadius": 4,
                  "borderTopRightRadius": 4,
                  "borderTopWidth": 1,
                  "marginLeft": 16,
                  "marginTop": 16,
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  [
                    {
                      "backgroundColor": "#e5e7eb",
                    },
                    {
                      "paddingBottom": 8,
                      "paddingLeft": 8,
                      "paddingRight": 8,
                      "paddingTop": 8,
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      {
                        "height": 140,
                        "width": 140,
                      },
                    ]
                  }
                >
                  <FastImageView
                    defaultSource={
                      {
                        "testUri": "../../../app/assets/images/pokeball.png",
                      }
                    }
                    resizeMode="cover"
                    style={
                      {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  />
                </View>
                <Text
                  style={
                    [
                      {
                        "color": "#1e293b",
                      },
                      {
                        "fontFamily": "Raleway-Regular",
                      },
                    ]
                  }
                >
                  pokemon
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
